src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -3844,3 +3849,21 @@ tapAsync f (Stream step1 state1) = Stream step TapInit
             Yield a s -> Yield a (TapDone s)
             Skip s    -> Skip (TapDone s)
             Stop      -> Stop
+
+-- XXX Is there room for improvement?
```

Comment:
If we make the ring buffer a first class data structure, we can just return the ring buffer instead of converting it into an array first. Consumers can directly consume the ring buffer.